File: src/scripts/backfill-interaction-embeddings.ts
/**
 * backfill-interaction-embeddings.ts
 *
 * Fills in the `interactionEmbedding` field for records in
 * `patron_interaction_history` that were created without an embedding
 * (e.g. by seed-interactions.ts).
 *
 * Sends records to the LiteLLM gateway → local TEI (voyage-4-nano) in batches
 * of up to BATCH_SIZE (128), with bounded concurrency (CONCURRENCY).
 *
 * Usage:
 *   npm run backfill:interactions
 *
 * Requirements:
 *   LITELLM_BASE_URL and LITELLM_API_KEY must be set in .env
 */

import dotenv from "dotenv";
dotenv.config();

import type { Collection } from "mongodb";
import { getDb, closeClient } from "../db.js";
import { collections } from "../modeling/indexes.js";
import {
  generateEmbeddingBatch,
  buildInteractionEmbeddingText,
} from "../web/llm/embeddings.js";
import { config } from "../config.js";
import type { PatronInteractionRecord } from "../types.js";

/** Voyage AI max inputs per request */
const BATCH_SIZE = 128;

/** Number of batches to run concurrently */
const CONCURRENCY = 3;

type RecordProjection = {
  _id: unknown;
  interactionId: string;
  type: string;
  totalValueHKD: number;
  patronTierAtTime: string;
  detail: Record<string, unknown>;
  occurredAt: Date;
};

async function processBatch(
  col: Collection<PatronInteractionRecord>,
  batch: RecordProjection[],
  batchIndex: number,
  totalBatches: number
): Promise<{ success: number; failed: number }> {
  const texts = batch.map((rec) =>
    buildInteractionEmbeddingText({
      type:          rec.type,
      totalValueHKD: rec.totalValueHKD,
      tier:          rec.patronTierAtTime,
      detail:        rec.detail ?? {},
      occurredAt:    rec.occurredAt,
    })
  );

  const embeddings = await generateEmbeddingBatch(texts, "document");

  let success = 0;
  let failed = 0;

  const writes = batch.map(async (rec, i) => {
    const embedding = embeddings[i];
    const isNonZero = embedding.some((v) => v !== 0);

    if (isNonZero) {
      await col.updateOne(
        { _id: rec._id as never },
        { $set: { interactionEmbedding: embedding } }
      );
      success++;
    } else {
      console.warn(`  ⚠ ${rec.interactionId} — zero embedding returned, skipped`);
      failed++;
    }
  });

  await Promise.all(writes);

  console.log(
    `  Batch ${batchIndex + 1}/${totalBatches} done — ${success} ok, ${failed} skipped`
  );

  return { success, failed };
}

async function backfillInteractionEmbeddings() {
  if (!config.llm.apiKey) {
    console.error("LITELLM_API_KEY is not set in .env — cannot generate embeddings.");
    process.exitCode = 1;
    return;
  }

  const db = await getDb();
  const col = db.collection<PatronInteractionRecord>(collections.patronInteractions);

  // Find all records missing an embedding
  const missing = await col
    .find({ interactionEmbedding: { $exists: false } })
    .project<RecordProjection>({
      _id: 1,
      interactionId: 1,
      type: 1,
      totalValueHKD: 1,
      patronTierAtTime: 1,
      detail: 1,
      occurredAt: 1,
    })
    .toArray();

  const total = missing.length;

  if (total === 0) {
    console.log("All interaction records already have embeddings. Nothing to do.");
    return;
  }

  // Split into batches of BATCH_SIZE
  const batches: RecordProjection[][] = [];
  for (let i = 0; i < missing.length; i += BATCH_SIZE) {
    batches.push(missing.slice(i, i + BATCH_SIZE));
  }
  const totalBatches = batches.length;

  console.log(`\nFound ${total} records without embeddings.`);
  console.log(`Batches: ${totalBatches} × up to ${BATCH_SIZE} records, concurrency ${CONCURRENCY}`);
  console.log(`Estimated time: a few seconds\n`);

  let totalSuccess = 0;
  let totalFailed = 0;

  // Process batches with bounded concurrency
  for (let i = 0; i < batches.length; i += CONCURRENCY) {
    const window = batches.slice(i, i + CONCURRENCY);
    const results = await Promise.all(
      window.map((batch, j) =>
        processBatch(col, batch, i + j, totalBatches)
      )
    );
    for (const r of results) {
      totalSuccess += r.success;
      totalFailed  += r.failed;
    }
  }

  console.log(
    `\nBackfill complete: ${totalSuccess} succeeded, ${totalFailed} failed out of ${total} records.`
  );

  if (totalSuccess > 0) {
    console.log(
      `\nThe Atlas Vector Search index "interaction_embedding_idx" will index the new embeddings automatically.`
    );
    console.log(`You can now use the KPI Vector Search in the PR Efficiency tab.\n`);
  }
}

backfillInteractionEmbeddings()
  .catch((err) => {
    console.error("backfill-interaction-embeddings failed:", err);
    process.exitCode = 1;
  })
  .finally(closeClient);

File: server/app/scripts/backfill_interaction_embeddings.py
"""Fill ``interactionEmbedding`` for ``patron_interaction_history``.

Reads every record missing ``interactionEmbedding``, batches them (128 per
voyage call, 3 batches in parallel), and writes back the embedding using
``build_interaction_embedding_text``.

Port of ``src/scripts/backfill-interaction-embeddings.ts``.

Run with::

    uv run python -m app.scripts.backfill_interaction_embeddings
"""

from __future__ import annotations

import asyncio
from typing import Any

from app.collections import web_collections as cols
from app.db import close_client, get_db
from app.llm.embeddings import build_interaction_embedding_text, generate_embedding_batch

BATCH_SIZE = 128
CONCURRENCY = 3


async def _process_batch(
    coll: Any,
    batch: list[dict[str, Any]],
    batch_index: int,
    total_batches: int,
) -> tuple[int, int]:
    texts = [
        build_interaction_embedding_text(
            type_=rec.get("type"),
            total_value_hkd=float(rec.get("totalValueHKD") or 0),
            tier=rec.get("patronTierAtTime"),
            detail=rec.get("detail") or {},
            occurred_at=rec.get("occurredAt"),
        )
        for rec in batch
    ]

    embeddings = await generate_embedding_batch(texts, "document")

    success = 0
    failed = 0

    async def _write(rec: dict[str, Any], embedding: list[float]) -> None:
        nonlocal success, failed
        is_non_zero = any(v != 0 for v in embedding)
        if is_non_zero:
            await coll.update_one(
                {"_id": rec["_id"]},
                {"$set": {"interactionEmbedding": embedding}},
            )
            success += 1
        else:
            print(
                f"  ⚠ {rec.get('interactionId')} — zero embedding returned, skipped"
            )
            failed += 1

    await asyncio.gather(
        *[_write(rec, emb) for rec, emb in zip(batch, embeddings, strict=False)]
    )

    print(
        f"  Batch {batch_index + 1}/{total_batches} done — "
        f"{success} ok, {failed} skipped"
    )

    return success, failed


async def main() -> None:
    db = get_db()
    coll = db[cols.patron_interactions]

    missing = await (
        coll.find(
            {"interactionEmbedding": {"$exists": False}},
            projection={
                "_id": 1,
                "interactionId": 1,
                "type": 1,
                "totalValueHKD": 1,
                "patronTierAtTime": 1,
                "detail": 1,
                "occurredAt": 1,
            },
        ).to_list(length=None)
    )

    total = len(missing)
    if total == 0:
        print("All interaction records already have embeddings. Nothing to do.")
        return

    batches: list[list[dict[str, Any]]] = [
        missing[i : i + BATCH_SIZE] for i in range(0, total, BATCH_SIZE)
    ]
    total_batches = len(batches)

    print(f"\nFound {total} records without embeddings.")
    print(
        f"Batches: {total_batches} × up to {BATCH_SIZE} records, "
        f"concurrency {CONCURRENCY}"
    )
    print("Estimated time: a few seconds\n")

    total_success = 0
    total_failed = 0

    for i in range(0, len(batches), CONCURRENCY):
        window = batches[i : i + CONCURRENCY]
        results = await asyncio.gather(
            *[
                _process_batch(coll, batch, i + j, total_batches)
                for j, batch in enumerate(window)
            ]
        )
        for s, f in results:
            total_success += s
            total_failed += f

    print(
        f"\nBackfill complete: {total_success} succeeded, "
        f"{total_failed} failed out of {total} records."
    )

    if total_success > 0:
        print(
            "\nThe Atlas Vector Search index \"interaction_embedding_idx\" will "
            "index the new embeddings automatically."
        )
        print("You can now use the KPI Vector Search in the PR Efficiency tab.\n")


async def _entrypoint() -> None:
    try:
        await main()
    finally:
        await close_client()


if __name__ == "__main__":
    asyncio.run(_entrypoint())
